Reset worker cycle iterator for determinism across runs

evren-cerebras opened the pull request:
Reset worker cycle iterator for determinism across runs

Fixes #73603

evren-cerebras committed:
Fix for Issue 73603: Reset worker cycle iterator for determinism across runs

evren-cerebras commented:
@[emcastillo](https://github.com/emcastillo)

ejguan reviewed (commented):
It seems we can remove https://github.com/pytorch/pytorch/blob/67a275c29338a6c6cc405bf143e63d53abe600bf/torch/utils/data/dataloader.py#L898 from `__init__` as it will invoke `_reset`.

And, could you please add a test regarding this determinism?

evren-cerebras committed:
add test, remove redundant line

evren-cerebras commented:
@ejguan I added a test and removed the redundant line.  I spent a couple hours trying to build the pytorch repo from src to test this but it didn't work.  Had to stop because of time limitations.  

At the end I was stuck on this error:
`[4112/6126] Building CXX object third_party/ideep/mkl-dnn/third_party/oneDNN/src/common/CMakeFiles/dnnl_common.dir/gemm.cpp.o
ninja: build stopped: subcommand failed.
`

So I have not been able to test this in the test framework but I ran the test without and without these changes in a different way.

ejguan commented on test/test_dataloader.py:
```diff
@@ -2605,6 +2605,41 @@ def test_ind_worker_queue(self):
+
+        dataset = TestMultiEpochDataset(batch_size*num_workers)
+
+        loader = DataLoader(
+            dataset, batch_size=batch_size, num_workers=num_workers,
+            shuffle=False,  persistent_workers=True, timeout=5
+        )
```
Could you please check both `True` and `False` for `persistent_workers`?

evren-cerebras replied:
it works with False.  it's a good check since we remove the iterator intialization in __init__

evren-cerebras replied:
added test for false

ejguan commented on test/test_dataloader.py:
```diff
@@ -2605,6 +2605,41 @@ def test_ind_worker_queue(self):
+        worker_info = torch.utils.data.get_worker_info()
+        assert worker_info is not None
+        worker_id = worker_info.id
+        assert self.length % worker_info.num_workers == 0
```
What if the `length` cannot be divided by `num_workers` without remainder, does this test still work?

evren-cerebras replied:
it does work.  we may even be able to do this without batching.  i was being careful to avoid mixing a batch with data from different workers.

ejguan replied:
Let me pull your PR later today to do an extra test.

evren-cerebras replied:
i pushed some updates

ejguan reviewed (commented):
You can install pytorch nightly for local test. Then, you don't need to worry about the compilation.

ejguan commented:
Please also fix the lint error. (flake8). I don't think clang-format Error comes from your PR.

evren-cerebras committed:
remove unneeded assert on length, add test for persistent_workers False, fix formatting issues

ejguan commented on test/test_dataloader.py:
```diff
@@ -2605,6 +2605,41 @@ def test_ind_worker_queue(self):
+        assert worker_info is not None
+        worker_id = worker_info.id
+        for idx in range(self.length // worker_info.num_workers):
+            if worker_id == 0:
+                time.sleep(0.001)
```
I don't think this is needed for the test. Even without this `sleep`, the DataLoader before this PR would break this test.
The Error should be non-deterministic before your PR because it depends on which worker send `StopIteration` first.

evren-cerebras replied:
This line made the test fail with high probability for me that's why I added it.  Without it, you can something run multiple epochs and multiple test runs and not hit the issue.  I'm ok removing it but it will make the probability of detecting the issue lower.  On a similar note I got a timeout error in the last test run, not sure why this would cause that but I can't figure out why the test took over 5 sec to run.

evren-cerebras replied:
I took it out so it's fine.  One of the CI runs on an earlier commit failed with timeout.  I don't know how to get those previous logs.

ejguan commented on test/test_dataloader.py:
```diff
@@ -2605,6 +2605,41 @@ def test_ind_worker_queue(self):
+        return self.length
+
+
+class TestMultiEpochDeterminism(TestCase):
+    def test_multi_epochs_deter(self):
+        num_workers = 2
+        batch_size = 10
+        num_epochs = 3
+
+        dataset = TestMultiEpochDataset(batch_size * num_workers)
+
+        for persistent in [True, False]:
+            loader = DataLoader(
+                dataset, batch_size=batch_size, num_workers=num_workers,
+                shuffle=False, persistent_workers=persistent, timeout=5
+            )
+            for ind in range(num_epochs):
+                for batch_idx, sample in enumerate(loader):
+                    self.assertEqual(sample.tolist(), [batch_idx % num_workers] * batch_size)
```
super nit: Sorry for the confusion. I mean you can add this test under https://github.com/pytorch/pytorch/blob/5167e9d59d76d67f4a240f94600ce6bba2c70ed3/test/test_dataloader.py#L868
and use `self._get_data_loader` to get the dataloader.
`TestDataLoader` is inherited by `TestDataLoaderPersistentWorkers` to carry out the same test with `persistent_worker=True`. https://github.com/pytorch/pytorch/blob/5167e9d59d76d67f4a240f94600ce6bba2c70ed3/test/test_dataloader.py#L2334

evren-cerebras replied:
sure i can move it under that test.

evren-cerebras replied:
@ejguan with the sleep removed the test will randomly pass over 50% of the time for me.  I remove the sleep line, but just a heads up that it can lead to passing tests.

Hopefully I got the other changes you requested too.  

Thanks!

ejguan reviewed (commented):
In general, lgtm. Adding two more nit comments below. I have tested several corner cases like unbalanced data from each worker. It seems the DataLoader works fine with your PR. Thanks for adding this patch.

evren-cerebras committed:
move test into TestDataLoader, remove sleep
